### PR TITLE
Compute cached

### PR DESCRIPTION
**Pull request by zvookin**
Opening a PR to get comments. This provides zero copy memoization. I will likely work on factoring the changes to StorageFlattening into a separate pass.

### CODE REVIEW

**Comment by steven-johnson** [comment]
Thought: we could reduce the overhead for calculating the hash value by precalculating the 'fixed' portion (toplevel name + key name), either at compile time (if we're willing to bake assumptions about the hash function) or once-at-runtime (on first call). Patching this in and trying it, I get cache keys of ~72 bytes, of which only the last ~8 bytes vary with input parameters. (Admittedly, saving the hash calculation of 64 bytes may be a grasping at optimizations...)

**Comment by zvookin** [comment]
This is ready to review. I think there may be one docs TODO, which I'll check on tomorrow. Also it would be great if others could look at the USE_FULL_NAMES_IN_KEY #ifdef in Memoization.cpp. It addresses the comment above from Steven by using a pointer to a constant string and a 32-bit counter to identify the filter and Func in the cache key instead of each of their names as strings (with lengths as well).

The PR should make Memoization much faster. It is techncially a breaking change in that one can no longer memoize an output. That now results in a compile time error. The fix is to insert a compute_root'ed Func before the output Func and memoize the inner Func instead. (There is a change to the memoize.cpp test which demonstrates this.)

There is some question as to whether separating out the rewrite_memoized_allocations pass from storage_flattening is a good idea. The rewrite ends up being very dependent on the IR that storage flattening generates.

**Comment by steven-johnson** [comment]
Something is wonky with the latest changes; I pulled into a local test and it seems to get caught in an endless loop. Investigating.

**Comment by steven-johnson** [comment]
Edit: it's not endless, but broken: with USE_FULL_NAMES_IN_KEY=1 my cache hit rate is very close to 100%; with USE_FULL_NAMES_IN_KEY=0 it's under 30%.

**Comment by steven-johnson** [comment]
Looks like this line:
         size_so_far = (size_so_far + needed_alignment) & ~(needed_alignment - 1);
should be:
         size_so_far = (size_so_far + needed_alignment - 1) & ~(needed_alignment - 1);

(As it is, we round up the allocation size more than we need and end up with uninitialized bytes in the cache key.)

**Comment by zvookin** [comment]
Fix pushed.

File: src/Memoization.cpp [hunk 1]
```diff
@@ -162,6 +163,17 @@ class KeyInfo {
...
+// This ifdef is temporary to discuss the tradeoff between the two
+// approaches in code review of the pull request.  Using the full
+// names in the key results in a (hopefulyl incredibly slight)
+// performance differnece based on how one names filters and
+// functions. It is arguably a little easier to debug if something
+// goes wrong as one doesn't need to destructure the cache key. Also,
+// if a pointer is used, a counter must also be put in the cache key
+// to avoid aliasing on reuse of the address in JIT situations where
+// code is regenerated into the same region of memory.
+#define USE_FULL_NAMES_IN_KEY 0
```
Comment (steven-johnson) [inline src/Memoization.cpp]: Obviously we need to decide on this before landing. On my pathological-ish test case it definitely moved the needle performance wise so I vote for setting it to zero.

File: src/runtime/cache.cpp [hunk 2]
```diff
@@ -343,15 +369,28 @@ WEAK bool halide_memoization_cache_lookup(void *user_context, const uint8_t *cac
 
                 for (int32_t i = 0; i < tuple_count; i++) {
                     buffer_t *buf = tuple_buffers[i];
-                    copy_from_to(user_context, entry->buffer(i), *buf);
+                    //copy_from_to(user_context, entry->buffer(i), *buf);
```
Comment (steven-johnson) [inline src/runtime/cache.cpp]: Why is this commented out? Explain or remove, methinks

File: src/runtime/cache.cpp [hunk 3]
```diff
@@ -343,15 +369,28 @@ WEAK bool halide_memoization_cache_lookup(void *user_context, const uint8_t *cac
...
                 return false;
             }
         }
         entry = entry->next;
     }
 
+    for (int32_t i = 0; i < tuple_count; i++) {
+        buffer_t *buf = tuple_buffers[i];
+        size_t buffer_size = full_extent(*buf);
+        // TODO: ERROR RETURN
```
Comment (steven-johnson) [inline src/runtime/cache.cpp]: Is this TODO still a thing? If so I think it needs some fleshing out

**Comment by zvookin** [comment]
I fixed the TODOs by adding memory allocation failure logic and a test to make sure it works. Only remaining thing I know of is to decide what to do about the USE_FULL_NAMES_IN_KEY ifdefs.

**Comment by steven-johnson** [comment]
> what to do about the USE_FULL_NAMES_IN_KEY

Normally I'm reluctant to leave ifdef'ed code in place, but since there's a measurable performance win for the =0 case and a meaningful debugging win for the =1 case, tempted to leave it there, defaulted to zero.

**Comment by zvookin** [comment]
I updated the USE_FULL_NAMES_IN_KEY comment per Steven's last reply. Barring more review feedback or test failures in broader testing, I think this is ready to go.

File: src/IR.h [hunk 4]
```diff
@@ -271,17 +271,25 @@ struct Allocate : public StmtNode<Allocate> {
     Type type;
     std::vector<Expr> extents;
     Expr condition;
+
+    // These override the code generator dependent malloc and freeequivalents
+    // if provided. If the new succeeds, the delete is guaranteed to be called.
+    // It is allowed for the delete_stmt to be undefined even if
```
Comment (abadams) [inline src/IR.h]: delete_stmt no longer exists

File: src/IR.h [hunk 5]
```diff
@@ -271,17 +271,25 @@ struct Allocate : public StmtNode<Allocate> {
     Type type;
     std::vector<Expr> extents;
     Expr condition;
+
+    // These override the code generator dependent malloc and freeequivalents
```
Comment (abadams) [inline src/IR.h]: freeequivalents

Comment (steven-johnson) [inline src/IR.h]: NEW WORD OF THE DAY

File: src/runtime/cache.cpp [hunk 6]
```diff
@@ -70,20 +70,16 @@ WEAK size_t full_extent(const buffer_t &buf) {
 }
 
 WEAK void copy_from_to(void *user_context, const buffer_t &from, buffer_t &to) {
```
Comment (abadams) [inline src/runtime/cache.cpp]: This function now seems misleadingly named.

File: src/runtime/cache.cpp [hunk 7]
```diff
@@ -70,20 +70,16 @@ WEAK size_t full_extent(const buffer_t &buf) {
...
     for (int i = 0; i < 4; i++) {
         halide_assert(user_context, from.extent[i] == to.extent[i]);
         halide_assert(user_context, from.stride[i] == to.stride[i]);
     }
-    memcpy(to.host, from.host, buffer_size * from.elem_size);
+    to.host = from.host;
 }
 
 WEAK buffer_t copy_of_buffer(void *user_context, const buffer_t &buf) {
     buffer_t result = buf;
```
Comment (abadams) [inline src/runtime/cache.cpp]: Is this function a noop? It looks like it just returns a shallow copy of the buffer_t now.

File: src/runtime/windows_clock.cpp [hunk 8]
```diff
@@ -35,7 +35,7 @@ WEAK int64_t halide_current_time_ns(void *user_context) {
 }
 
 WEAK void halide_sleep_ms(void *user_context, int ms) {
-	Sleep(ms);
+        Sleep(ms);
```
Comment (abadams) [inline src/runtime/windows_clock.cpp]: Looks like I introduced these tabs via msvc, and this PR changes them to 8 spaces. Mind doing a drive-by conversion to 4 spaces in this PR? (to avoid the merge conflict if I fix in master)

**Comment by abadams** [comment]
LGTM. Will test on Windows and report back.

**Comment by abadams** [comment]
Crashes randomly on win32. Turning on using full names in key seems to fix it.

**Comment by abadams** [comment]
Good news! Also crashes randomly on linux in 32-bit mode. That should make the debugging easier.

**Comment by steven-johnson** [comment]
Hmm, I think Handle().bytes() is always 64 bits, even for 32-bit mode... in which case we could have uninitialized bytes in the upper (lower?) half.

**Comment by abadams** [comment]
Yeah, that looked like the issue. The fix was non-trivial, so I made it a PR into master ( #880 )
